src/Video/ParticleSystemRenderer.hpp

```diff
@@ -0,0 +1,134 @@
+    glm::mat4 m = glm::mat4(1);
+};
+
+namespace Video {
+    class Texture;
+
+    class ParticleSystemRenderer {
+        public:
+        struct EmitterSettings {
+            /// Number of particles.
+            int nr_particles = 1024;
+
+            /// Position.
+            glm::vec3 worldPos;
+
+            /// Size.
+            glm::vec2 size;
+
+            /// Life (in seconds).
+            float life;
+
+            /// Lifetime (in seconds).
+            float lifetime = 10;
+
+            /// Initial velocity.
+            glm::vec3 velocity;
+
+            /// Start, mid and end of life alpha of particle.
+            glm::vec3 alpha;
+
+            /// Color of the particle.
+            glm::vec3 color = glm::vec3(0, 0, 0);
+
+            /// Fire rate.
+            float rate = 1.0f;
+
+            /// Texture index (for the texture atlas, left to right, top to bottom indexing)
+            int textureIndex = 0;
+
+            /// Multiply velocity.
+            float velocityMultiplier = 10.0f;
+
+            /// Gives spread to the particles
+            int spread = 1;
+
+            /// Used to give random velocity to particles at all time.
+            glm::vec3 randomVec = glm::vec3(1, 1, 1);
+
+            /// Scale.
+            float scale = 1.0f;
+
+            /// Particle mass.
+            float mass = 0.01f;
+
+            /// Alpha control
+            float alpha_control = 1.0f;
+
+            /// Nr Of new particles per emitt, max is 32.
+            int nr_new_particles = 31;
+        };
+
+        EmitterSettings emitterSettings;
+
+        VIDEO_API ParticleSystemRenderer();
+        VIDEO_API ParticleSystemRenderer(int count);
+        VIDEO_API ~ParticleSystemRenderer();
+
+        VIDEO_API void Init();
+
+        // Creates storage buffers for particles.
+        VIDEO_API void CreateStorageBuffers();
+
+        // Particles are sent to the compute shader and we compute there the new positions/velocities.
+        VIDEO_API void Update(float dt, EmitterSettings settings);
+
+        //Render the particles.
+        VIDEO_API void Draw(Texture* textureAtlas, unsigned int textureAtlasRows, const glm::mat4& viewProjectionMatrix, ParticleSystemRenderer::EmitterSettings settings);
+
+        private:
+        Video::ShaderProgram* computeShaderProgram;
+
+        Video::ShaderProgram* shaderProgram;
+
+        int WORK_GROUP_SIZE = 128;
+
+        unsigned int nr_particles = 1024 * 32;
+        glm::vec2 particleShootIndex = glm::vec2(0, 30);
+        int nr_new_particles = 31;
+        float rate = 1000.0f;
+        float delta_time = 0.0f;
+        ParticlePos* points;
+        ParticleVelocity* vels;
+        ParticleColor* col;
+        ParticleModelMatrix* rots;
+        glm::mat4 rotMat;
+
+        float timer = 0.0f;
+
+        GLuint posSSbo;
+        GLuint velSSbo;
+        GLuint colSSbo;
+        GLuint mmSSbo;
+        GLuint m_glDrawVAO;
+    };
+} // namespace Video
```

Comment:
Missing newline at end of file.

Reply:
There is a newline at the end of this file.